pkg/steps/bundle_source.go

```diff
@@ -69,58 +69,49 @@ func (s *bundleSourceStep) run(ctx context.Context) error {
 	return handleBuild(ctx, s.buildClient, build, s.artifactDir)
 }
 
-func replaceCommand(manifestDir, pullSpec, with string) string {
-	return fmt.Sprintf("find %s -type f -exec sed -i 's?%s?%s?g' {} +", manifestDir, pullSpec, with)
+func replaceCommand(pullSpec, with string) string {
+	return fmt.Sprintf("find . -type f -name \\*.yaml -exec sed -i 's?%s?%s?g' {} +", pullSpec, with)
 }
 
 func (s *bundleSourceStep) bundleSourceDockerfile() (string, error) {
 	var dockerCommands []string
 	dockerCommands = append(dockerCommands, "")
 	dockerCommands = append(dockerCommands, fmt.Sprintf("FROM %s:%s", api.PipelineImageStream, api.PipelineImageStreamTagReferenceSource))
-	manifestDir := filepath.Join(s.config.ContextDir, s.config.OperatorManifests)
-	for _, sub := range s.config.Substitute {
-		replaceSpec, err := s.getFullPullSpec(sub.With)
-		if err != nil {
+	is, err := s.imageClient.ImageStreams(s.jobSpec.Namespace()).Get(context.TODO(), api.StableImageStream, meta.GetOptions{})
+	if err != nil {
+		return "", fmt.Errorf("failed to get stable imagestream: %w", err)
+	}
+	for _, sub := range s.config.Substitutions {
+		replaceSpec, exists := util.ResolvePullSpec(is, sub.With, false)
+		if !exists {
 			return "", fmt.Errorf("failed to get replacement imagestream for image tag `%s`", sub.With)
 		}
-		dockerCommands = append(dockerCommands, fmt.Sprintf(`RUN ["bash", "-c", "%s"]`, replaceCommand(manifestDir, sub.PullSpec, replaceSpec)))
+		// The \ character has to be escaped in the dockerfile to run correctly
+		replace := strings.ReplaceAll(replaceCommand(sub.PullSpec, replaceSpec), `\`, `\\`)
+		dockerCommands = append(dockerCommands, fmt.Sprintf(`RUN ["bash", "-c", "%s"]`, replace))
 	}
 	dockerCommands = append(dockerCommands, "")
 	return strings.Join(dockerCommands, "\n"), nil
 }
 
-func (s *bundleSourceStep) getFullPullSpec(tag string) (string, error) {
-	is, err := s.imageClient.ImageStreams(s.jobSpec.Namespace()).Get(context.TODO(), api.StableImageStream, meta.GetOptions{})
-	if err != nil {
-		return "", err
-	}
-	if len(is.Status.PublicDockerImageRepository) > 0 {
-		return is.Status.PublicDockerImageRepository + ":" + tag, nil
-	}
-	if len(is.Status.DockerImageRepository) > 0 {
-		return is.Status.DockerImageRepository + ":" + tag, nil
-	}
-	return "", fmt.Errorf("no pull spec available for image stream %s", api.StableImageStream)
-}
-
 func (s *bundleSourceStep) Requires() []api.StepLink {
 	return []api.StepLink{
 		api.InternalImageLink(api.PipelineImageStreamTagReferenceSource),
```

Comment:
This is problematic - we need to parse the substitutions to ensure we correctly expose images that we require to be present. If a user wants to substitute with a pull-spec from `pipeline` it must be built before this runs.

Reply:
I don't think this is that problematic - the images corresponding to the pullspecs only need to be available once you actually use the bundle - after you deploy the CatalogSource and create a `Subscription`, etc, but you don't need it at the bundle build time. So as long as we validate that the image names in `with` stanzas `will` resolve to somehting, we should be fine.